libraries/AP_RangeFinder/AP_RangeFinder_TeraRangerDUOSerial.cpp

```diff
@@ -0,0 +1,147 @@
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_TeraRangerDUOSerial.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <AP_Math/crc.h>
+#include <ctype.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+/* 
+   The constructor also initialises the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_TeraRangerDUOSerial::AP_RangeFinder_TeraRangerDUOSerial(RangeFinder::RangeFinder_State &_state,
+                                                           AP_RangeFinder_Params &_params,
+                                                           AP_SerialManager &serial_manager,
+                                                           uint8_t serial_instance) :
+    AP_RangeFinder_Backend(_state, _params)
+{
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+// detect if a TeraRanger DUO rangefinder sensor is connected by looking for a configured serial port
+bool AP_RangeFinder_TeraRangerDUOSerial::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// update the state of the sensor
+void AP_RangeFinder_TeraRangerDUOSerial::update(void)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    // process incoming messages
+    if (get_reading(state.distance_cm)) {
+        _last_reading_ms = AP_HAL::millis();        
+    }
+    update_status();
+}
+
+// update status
+void AP_RangeFinder_TeraRangerDUOSerial::update_status()
+{
+    if (AP_HAL::millis() - _last_reading_ms > TRDUO_TIMEOUT_MS) {
+        set_status(RangeFinder::RangeFinder_NoData);
+    } else {
+        if ((int16_t)state.distance_cm > TRDUO_MAX_DISTANCE_TOF) {
+            set_status(RangeFinder::RangeFinder_OutOfRangeHigh);
+        } else if ((int16_t)state.distance_cm < TRDUO_MIN_DISTANCE_SOUND) {
+            set_status(RangeFinder::RangeFinder_OutOfRangeLow);
+        } else {
+            set_status(RangeFinder::RangeFinder_Good);
+        }
+    }
+}
+
+// read - return last value measured by sensor
+bool AP_RangeFinder_TeraRangerDUOSerial::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_tof = 0;
+    float sum_sound = 0;
+    uint16_t tof_count = 0;
+    uint16_t sound_count = 0;
+    int16_t nbytes = uart->available();
+
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        if (c == 'T') {
+            _buffer_count = 0;
+            _found_start = true;
+        }
+
+        if (!_found_start) {
+            continue;
+        }
+
+        _buffer[_buffer_count++] = c;
```

Comment:
When the method returns a value of 1, it is added to the buffer if it is at the beginning.
It is a CRC check that detects errors.